### ISSUE TITLE
Recursive dependencies

### ISSUE THREAD

**Issue by MidnightDesign** [composer/composer#5019]
I've got two packages `vendor/foo` and `vendor/bar`. `foo` requires `bar` and vice-versa:

`vendor/foo`'s `composer.json`:
```json
{
    "name": "vendor/foo",
    "require": { "vendor/bar": "^1.0" },
}
```

`vendor/bar`'s `composer.json`:
```json
{
    "name": "vendor/bar",
    "require": { "vendor/foo": "^1.0" },
}
```

They are both in VCS repositories in Bitbucket. (I've omitted the `repositories` keys for brevity.)

Aside from the fact that there's probably something wrong with my dependencies, I get this weird behavior: if I try to `composer update` one of the packages (in the following case `bar`), I get something like

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - vendor/foo 1.0.0 requires eventjet/bar ^1.0.0 -> no matching package found.
    - Installation request for vendor/foo ^1.0 -> satisfiable by vendor/foo[1.0.0].
```

So it can't find _itself_. If I add a `"version": "1.0.0"` to `vendor/bar`'s composer.json, it works. Is this behavior expected? Is there any way around it?

Again: I know this is stupid. But that doesn't mean it shouldn't work, right?

**Comment by alcohol** [comment]
Circular dependencies do not make much sense.
Do both repositories contain a 1.0.0 git tag?

**Comment by MidnightDesign** [comment]
Yes, they both do have a 1.0.0 tag.

> Circular dependencies do not make much sense.

See the last sentence of my post. : )

**Comment by Seldaek** [comment]
It should work yes, and AFAIK it does, but it's impossible to see if you're doing it wrong as you omitted half the details. We need a full composer.json and exact commands you are running and ideally full output. If it depends on private packages try to reproduce it with public ones.

**Comment by MidnightDesign** [comment]
I wanted to spare you reading tons of stuff that doesn't relate to the problem. I'll try to reproduce the problem with public packages in the next couple of days.

**Comment by alcohol** [comment]
Even better if you can reproduce it as a test scenario by creating a fixture in `tests/Composer/Test/Fixtures/installer/`, that way we can avoid regressions.

**Comment by curry684** [comment]
What appears to be happening, guessing a bit from your example, is that you checked out branch `master` of `vendor/bar` and then run `c update`. The version guesser, not finding a `version` tag in `composer.json`, works from the current git branch, and for `master` resorts to `999999-dev`.

So the solver is completely correct - `vendor/bar:999999-dev` is already loaded and the only available solution requires `vendor/bar:^1.0` which doesn't match.

We could solve this neatly by ignoring version checks on circular requirements to the root package, would be functionally correct during development, but also a ton of work for a rather freakish edge case.

**Comment by Seldaek** [comment]
Yup I think the solution here is to use COMPOSER_ROOT_VERSION=1.0.0 to define the version if it can not be detected correctly, or add a [branch-alias](https://getcomposer.org/doc/articles/aliases.md#branch-alias) to let master resolve to 1.x
